pkgs/build-support/rust/build-rust-package/default.nix

```diff
@@ -15,145 +15,149 @@
+        } // lib.optionalAttrs (finalAttrs ? cargoHash) {
+          hash = finalAttrs.cargoHash;
+        } // lib.optionalAttrs (finalAttrs ? cargoSha256) {
+          sha256 = finalAttrs.cargoSha256;
+        } // (args.depsExtraArgs or { }));
+
+    maybeSetStr = x: lib.optionalString (previousAttrs ? ${x}) previousAttrs.${x};
+    maybeSetListFromPrevious = x: lib.optionals (previousAttrs ? ${x}) previousAttrs.${x};
+
+    # cargoBuildFeatures example:
+    # previous: use directly specified cargoBuildFeatures from the original derivation (the one in nixpkgs)
+    # final: use buildType from either the original derivation or one overriden with overrideAttrs
+    # default: use the default passed if none of the above are set
+    pickPreviousOrFinalOrDefault = previous: final: default: (previousAttrs.${previous} or finalAttrs.${final} or default);
+
+  in
+
+  # Tests don't currently work for `no_std`, and all custom sysroots are currently built without `std`.
+    # See https://os.phil-opp.com/testing/ for more information.
+  assert useSysroot -> !(finalAttrs.doCheck or true);
+  {
+    inherit buildAndTestSubdir cargoDeps;
+
+    # buildType is checked for 4 of the cargo built in types
+    # directly specifying cargoBuildType can be used as a escape-hatch
+    cargoBuildType = assert finalAttrs ? buildType -> lib.assertOneOf "buildType" finalAttrs.buildType [ "release" "debug" "test" "bench" ];
+      pickPreviousOrFinalOrDefault "cargoBuildType" "buildType" "release";
+
+    cargoCheckType = pickPreviousOrFinalOrDefault "cargoCheckType" "checkType" finalAttrs.cargoBuildType;
+
+    cargoBuildNoDefaultFeatures = pickPreviousOrFinalOrDefault "cargoBuildNoDefaultFeatures" "buildNoDefaultFeatures" false;
+
+    cargoCheckNoDefaultFeatures = pickPreviousOrFinalOrDefault "cargoCheckNoDefaultFeatures" "checkNoDefaultFeatures"
+      finalAttrs.cargoBuildNoDefaultFeatures;
+
+    cargoBuildFeatures = pickPreviousOrFinalOrDefault "cargoBuildFeatures" "buildFeatures" "";
+
+    cargoCheckFeatures = pickPreviousOrFinalOrDefault "cargoCheckFeatures" "checkFeatures" finalAttrs.cargoBuildFeatures;
+
+    patchRegistryDeps = ./patch-registry-deps;
+
+    nativeBuildInputs = maybeSetListFromPrevious "nativeBuildInputs" ++ [
+      cacert
+      git
+      cargoBuildHook
+      cargoCheckHook
+      cargoInstallHook
+      cargoSetupHook
+      rustc
+    ];
+
+    buildInputs = maybeSetListFromPrevious "buildInputs"
     ++ lib.optionals stdenv.hostPlatform.isDarwin [ libiconv ]
     ++ lib.optionals stdenv.hostPlatform.isMinGW [ windows.pthreads ];
 
-  patches = cargoPatches ++ patches;
+    # commented for now to keep drv hash from changing
+    #cargoPatches = maybeSetListFromPrevious "cargoPatches";
+    #patches = finalAttrs.cargoPatches ++ maybeSetListFromPrevious "patches";
+
+    patches = (finalAttrs.cargoPatches or [ ]) ++ maybeSetListFromPrevious "patches";
+
+    PKG_CONFIG_ALLOW_CROSS = previousAttrs.PKG_CONFIG_ALLOW_CROSS or
```

Comment:
does this (technically) change the behavior of this binding? looks like previously the value from `args` would always be overwritten.